### ISSUE TITLE
[unstable option] imports_layout

### ISSUE THREAD

**Issue by scampi** [rust-lang/rustfmt#3361]
Tracking issue for unstable option: imports_layout

**Comment by jimmycuadra** [comment]
What's necessary to stabilize this?

**Comment by scampi** [comment]
There are some steps described in https://github.com/rust-lang/rustfmt/blob/master/Processes.md#stabilising-an-option.
https://github.com/rust-lang/rustfmt/pull/3581 is an attempt at this.

**Comment by mzabaluev** [comment]
Why does `Horizontal` have to exceed `max_width`?
How does this behavior interact with `merge_imports = true`?

I think it should be possible to try to fit all import statements into `max_width`, replicating the path prefix  for any spill-over items into the next line(s).

**Comment by mzabaluev** [comment]
One good reason to prefer the `Horizontal` layout (and stabilize this setting after the behavioral aspects are ironed out) is to keep the imports searchable with grep-like tools that don't easily match across multiple lines.

**Comment by andreastt** [comment]
A counter-argument for `Vertical` sorting is that it makes rebases significantly easier.

I recently had to do a rebase of a fairly complicated refactoring that moved lots of types around, and git rebase tools for in-line changes is not great.

**Comment by vn971** [comment]
I would *love* independent imports:
```
use foo::{
    aaaaaaaaaaaaaaaaaa, bbbbbbbbbbbbbbbbbb, cccccccccccccccccc, dddddddddddddddddd,
    eeeeeeeeeeeeeeeeee, ffffffffffffffffff,
};
```
=>
```
use foo::aaaaaaaaaaaaaaaaaa;
use foo::bbbbbbbbbbbbbbbbbb;
use foo::cccccccccccccccccc;
use foo::dddddddddddddddddd;
use foo::eeeeeeeeeeeeeeeeee;
use foo::ffffffffffffffffff;
```

As a possible alternative, I find this style super-friendly to git, immediately readable to my tastes, easy to play around with commenting-uncommenting.

Is it possible to have it as an ~~independent~~ alternative option?

UPDATE: for unstable rustfmt, this has been implemented as [Item](https://rust-lang.github.io/rustfmt/?version=master&search=#Item%5C%3A) granularity :tada:

**Comment by dtolnay** [comment]
I think "independent imports" (https://github.com/rust-lang/rustfmt/issues/3361#issuecomment-632661079) is outside of what this option is for, and is a discussion for https://github.com/rust-lang/rustfmt/issues/3362 instead. It's already been brought up there in https://github.com/rust-lang/rustfmt/issues/3362#issuecomment-641125438.

AIUI the two options are orthogonal and the distinction is as follows: you can think of `merge_imports` as determining the *tokens* of our imports i.e. how/whether they should be grouped i.e. which parts of which specific paths should go inside of the same curly braces; while `imports_layout` determines the *whitespace* of each import after the point that the *tokens* have already been decided.

**Comment by Lokathor** [comment]
Shouldn't "mixed' be called "compressed" to match other similar options?

**Comment by vn971** [comment]
To explain for new people reading this issue. `imports_layout` is not going to be stabilized and is instead replaced with `imports_granularity` https://github.com/rust-lang/rustfmt/issues/3362#issuecomment-766415850

This issue can probably be even closed?

**Comment by calebcartwright** [comment]
> To explain for new people reading this issue. imports_layout is not going to be stabilized and is instead replaced with imports_granularity #3362 (comment)

This is incorrect. `merge_imports` was superseded by `imports_granularity`, there was no impact and no change to `imports_layout` which was a different option that controlled different formatting aspects.

**Comment by vn971** [comment]
@calebcartwright sorry, you're right! I've deleted my comment.

**Comment by Fishrock123** [comment]
Is this the same thing as `imports_granularity`? What kind of timeline / issues are there for stabilizing that?

**Comment by bartlomieju** [comment]
Deno project is very interested in this setting as well (because we only use latests stable Rust toolchain). 

We'd prefer to format to have a single import per line (which is is very git diff friendly).

**Comment by 9999years** [comment]
> Is this the same thing as `imports_granularity`? What kind of timeline / issues are there for stabilizing that?

I've opened #4991 to track `imports_granularity`.

**Comment by ytmimi** [comment]
linking #3286

**Comment by blyxyas** [comment]
The option `LimitedHorizontalVertical` [isn't documented](https://rust-lang.github.io/rustfmt/?version=v1.5.1&search=#imports_layout)

**Comment by ytmimi** [comment]
@blyxyas PRs for doc improvements are always welcome 😁

**Comment by blyxyas** [comment]
I haven't made a PR because I don't know what it does
Sorry

**Comment by ytmimi** [comment]
@blyxyas no worries. I also couldn't have told you what it did before looking into it. From what I can tell it's meant to let you control the width at which imports wrap from being laid out horizontally to being laid out vertically, however I don't think we support parsing config options that take a value, and that might be why its not documented. 

`imports_layout` config option maps to the `ListTactic` enum.
https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/config/mod.rs#L90
https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/config/lists.rs#L26-L41

I tried setting it from the command line and got an error

```
rustfmt --config="imports_layout=LimitedHorizontalVertical" 
invalid key=val pair: `imports_layout=LimitedHorizontalVertical`

rustfmt --config="imports_layout=LimitedHorizontalVertical(50)"
invalid key=val pair: `imports_layout=LimitedHorizontalVertical(50)`
```

The `ListTactic` enum is pretty general and is used internally to format many different types of lists. Maybe it makes more sense to create a new `ImportsListTactic` enum that exposes the same options as `ListTactic` except the `LimitedHorizontalVertical(50)` variant. Given that `LimitedHorizontalVertical(_)` can't be set (at least not from the command line) maybe it's not worth it to make that change.

**Comment by tgross35** [comment]
> If there are any nested imports in a list import, then use the multi-line form, even if the import fits on one line. Each nested import must be on its own line, but non-nested imports must be grouped on as few lines as possible.

I guess that does line up. Though I'm not sure why "Horizontal" breaks into a multiline group when it's not close to line length

**Comment by ytmimi** [comment]
> Though I'm not sure why "Horizontal" breaks into a multiline group when it's not close to line length

@tgross35  [From the style guide](https://github.com/rust-lang/style-team/blob/master/guide/items.md#nested-imports:~:text=If%20there%20are%20any%20nested%20imports%20in%20a%20list%20import%2C%20then%20use%20the%20multi%2Dline%20form%2C%20even%20if%20the%20import%20fits%20on%20one%20line.):
"If there are any nested imports in a list import, then use the multi-line form, even if the import fits on one line."

**Comment by tgross35** [comment]
It was just my initial understanding that `Horizontal` breaks the rules to keep everything on one line, e.g., in the docs it says that it may exceed line length. There just isn't an example with nested modules in the docs. (fwiw, I don't have any problem with the behavior, it just didn't line up with my impression from the docs)

**Comment by daniel-wong-dfinity-org** [comment]
Can we make this stable as-is? Otherwise, it seems like perfect is being made the enemy of good.

**Comment by calebcartwright** [comment]
> Can we make this stable as-is? 

No.

> Otherwise, it seems like perfect is being made the enemy of good.

That's not the case at all. There's a well documented set of criteria that our config options have to meet before they can be stabilized, not to mention that the Rust ecosystem takes the notion of stability pretty seriously. We don't stabilize something that has known bugs and known issues in interactions with other parts of the ecosystem (in this case other Rustfmt options)

https://github.com/rust-lang/rustfmt/discussions/5365

**Comment by daniel-wong-dfinity-org** [comment]
There are people who cannot use this even though it would provide value.

**Comment by boomshroom** [comment]
Apparently Linus decrees that [thou shalt use `Vertical` in the Linux kernel](https://lore.kernel.org/dri-devel/CAHk-=wgO7S_FZUSBbngG5vtejWOpzDfTTBkVvP3_yjJmFddbzA@mail.gmail.com/). So this, and #4991, would need to be stabilised for Rust code to have a better chance of getting merged without one of his signature rants.

**Comment by ojeda** [comment]
> Apparently Linus decrees that [thou shalt use `Vertical` in the Linux kernel](https://lore.kernel.org/dri-devel/CAHk-=wgO7S_FZUSBbngG5vtejWOpzDfTTBkVvP3_yjJmFddbzA@mail.gmail.com/). So this, and [#4991](https://github.com/rust-lang/rustfmt/issues/4991), would need to be stabilised for Rust code to have a better chance of getting merged without one of his signature rants.

Not necessarily -- `Vertical` is closer to what we want, but it may not be exactly it. For instance, it currently reformats single imports into a single line, and we may want to avoid that since they are useful either for consistency (i.e. always done like that) or at least when one knows more items will be added. Orthogonally, another approach that was brought up was using the existence of the trailing comma to influence the formatting (either to reformat when not in vertical layout or to accept the multiline formatting).

**Comment by pfsavm** [comment]
What does it need to finally bring this to `stable`?

We would like to use this in production and nightly is always a hurdle.
